### DISCUSSION TITLE
Webhooks No Longer Triggering after v3.7.0 Upgrade/Event Rule Migration

### DISCUSSION THREAD

**Discussion by eramnes** [netbox-community/netbox#14764] [Help Wanted!]
We have been using Netbox for a while now and had set up some webhooks to call some APIs to perform actions on object updates.  These had been working great.

We recently updated to v3.7.0 and with the migration to Event Rules, it seems the webhooks no longer trigger on object update.  The auto-creation of the Event Rules from the conditions on the old webhooks was successful, and the conditions are still there, but changing an object so that it would trigger the Event Rule does nothing.

I've checked the request queue after making a change that should match an Event Rule, and there's no request that shows up there, in any state (active, finished, failed, scheduled).  I've verified that netbox-rq is running, and the PIDs for the workers show in the request queue interface.  I also tried creating a brand new Event Rule with the same conditions and pointing it to the old webhook but that didn't seem to make a difference.  We are using custom Status values as the condition for the Event Rule, but I tried using a provided value instead and that didn't seem to help.

Given that I haven't seen anyone else report this as an issue, I think I must be doing something wrong.  Can anyone point me in the correct direction?  I've tried turning logging on for Netbox and Django but I don't see anything that stands out as having to do with webhooks or Event Rules.  If you need any further info I would be happy to provide that.

Thanks.

**Comment by DanSheps** [comment]
Could you provide the rule and webhook (sanitize as required obvivously)?

I do have a webhook (that does nothing) and that appears to be working as intended

**Reply by markkuleinio** [reply]
Works here as well: tcpdump shows all attempts to connect, and Failed Jobs show all my attempts: create, update, delete.

**Comment by eramnes** [comment]
Here is the event rule:
![event_rule](https://github.com/netbox-community/netbox/assets/8175464/c8a05645-7125-4f75-9769-86e3fef3e0a8)

Here is the webhook that the event rule should be calling:
![webhook](https://github.com/netbox-community/netbox/assets/8175464/31ff5d1b-a5fe-4feb-b8ab-b8af04f9eccb)

If there's a better way to provide the settings than screenshots please let me know.  Thanks!

**Reply by candlerb** [reply]
Is it *attempting* to make the connection? Go to `/admin/background-tasks/`, and check whether "Finished Jobs" or "Failed Jobs" increments.

If the failed or finished job counter increments, then click on it and then click on the job UUID to see the details of what happened.

If nothing increments, then I would next suspect the webhook condition. Try setting it to null (i.e. so it is triggered on all changes of VM). If it then works, then we can dig further into the condition. If it still does nothing, then other digging will be required.

Aside: it seems odd that your event rule only considers an *update* of a virtualmachine in Netbox, but you are sending a message that looks like "create VM" to the target.

**Reply by eramnes** [reply]
Test 1: Event Rule condition in place, virtual-machine object changed to "specced"

The webhook was not triggered.  I checked the background tasks and neither counter had incremented.  The virtual-machine status was reset to "active" prior to the next test.

Test 2: Event Rule condition removed, virtual-machine object changed to "specced"

The webhook was not triggered.  Again, there was no increment of the background tasks counter.  The virtual-machine status was reset to "active"

![queues](https://github.com/netbox-community/netbox/assets/8175464/bb6161bc-b878-4ab3-a597-7b52eeedf1e3)
This is the view of the background-tasks page after both tests.

Aside: It is weird logic, but it has to do with the requirement to add an Interface to the VM in Netbox before creating it so that the Netbox IPAM pieces are updated correctly from our external DNS server.  If the interface isn't there it doesn't know what VRF or VLAN to use.  These are provided by the Interface but the VM has to exist first.  I'm guessing there's probably a better way to do this but I haven't done any work on it lately.

**Reply by candlerb** [reply]
> Test 2: Event Rule condition removed, virtual-machine object changed to "specced"
> 
> The webhook was not triggered. Again, there was no increment of the background tasks counter. The virtual-machine status was reset to "active"

That is suspicious; the whole database transaction seems to have been rolled back. How are you changing the status - just normally in the web UI? Does `journalctl -eu netbox` show any errors? What happens if you remove the event rule entirely? I wonder if you have a completely separate issue with virtualmachines and your custom status values, which is unrelated to event rules / webhooks.

Webhooks are working for me on 3.7.0, but the object type of interest is IPAddress, and I don't have any filter in the event rule.

**Reply by markkuleinio** [reply]
I tested both prefix and VM, all works for me on 3.7.0 (no conditions)

**Reply by eramnes** [reply]
The status is being changed via the web UI - I can try to change via API if you think that would be worth testing.

There doesn't appear to be anything in journalctl that's related at least starting from the most recent services restart (Central Time, sanitized):

[user@netbox ~]$ sudo journalctl -eu netbox
Jan 10 10:50:24 netbox.example.com systemd[1]: Started NetBox WSGI Service.
Jan 10 10:50:24 netbox.example.com gunicorn[43193]: [2024-01-10 10:50:24 -0600] [43193] [INFO] Starting gunicorn 21.2.0
Jan 10 10:50:24 netbox.example.com gunicorn[43193]: [2024-01-10 10:50:24 -0600] [43193] [INFO] Listening at: http://127.0.0.1:8001 (43193)
Jan 10 10:50:24 netbox.example.com gunicorn[43193]: [2024-01-10 10:50:24 -0600] [43193] [INFO] Using worker: gthread
Jan 10 10:50:24 netbox.example.com gunicorn[43194]: [2024-01-10 10:50:24 -0600] [43194] [INFO] Booting worker with pid: 43194
Jan 10 10:50:24 netbox.example.com gunicorn[43195]: [2024-01-10 10:50:24 -0600] [43195] [INFO] Booting worker with pid: 43195
Jan 10 10:50:24 netbox.example.com gunicorn[43196]: [2024-01-10 10:50:24 -0600] [43196] [INFO] Booting worker with pid: 43196
Jan 10 10:50:24 netbox.example.com gunicorn[43197]: [2024-01-10 10:50:24 -0600] [43197] [INFO] Booting worker with pid: 43197
Jan 10 10:50:24 netbox.example.com gunicorn[43198]: [2024-01-10 10:50:24 -0600] [43198] [INFO] Booting worker with pid: 43198

If I remove the event rule, should I recreate it, or is there some other test to do after it's gone?

I can pull the custom status values out of configuration.py if you think that might be an issue, but I did previously try changing the event rule to use a delivered status ("planned") and it didn't trigger on that either.  The custom status values existed prior to the 3.7.0 upgrade.

**Reply by candlerb** [reply]
> The status is being changed via the web UI - I can try to change via API if you think that would be worth testing.

No: I just wanted to be sure it wasn't something odd like a Custom Script.

If you change the status, it should persist. If it silently reverts to the original status when you click Save (which is what I think you're saying), that's borked. But if it raises an exception which rolls back the transaction, you ought to have got an exception in the GUI too.  It [makes no sense](https://en.wikipedia.org/wiki/Chewbacca_defense).

> If I remove the event rule, should I recreate it, or is there some other test to do after it's gone?

Remove the event rule, make the status change through the GUI, see if it silently reverts to the old value. If it does, you've completely taken event rules out of the loop and you know the problem lies elsewhere.

If the status changes successfully, then put the event rule back. Now if you change the status through the GUI, does it still persist changes, or does it reset?

Custom statuses might or might not be an issue (and as you've said, you can remove them to rule them out). I wonder if there's some issue with serialization of your object into the event queue, but I still don't get why the problem is silent.

**Reply by eramnes** [reply]
I think I might not have explained properly what's happening when the virtual-machine status changes.  If the status is set to "active" as an example, and I change it to "specced" in the web UI, the change occurs and the status does stick around after the change:

![specced](https://github.com/netbox-community/netbox/assets/8175464/ce3622bd-21bb-416b-9ea4-e9a66c6f86a7)
![specced_cl](https://github.com/netbox-community/netbox/assets/8175464/d637626a-e8db-4b82-a814-7c3713f30d2d)

So it's not that the status change doesn't stick around, it's that the value of "specced" and the operation of "updated" aren't triggering the event rule even though it seems like it should.

I do not see any errors when modifying status, either in the web UI or in the background-tasks page.  I am not sure if it's logging something where I'm unable to see it.

**Reply by candlerb** [reply]
It seems I misunderstood what you meant by *The virtual-machine status was reset to "active"*

If the changelog is showing the save, and you also tried it without any event filter, then I don't know what's happening. It certainly works for me (when updating an IP Address, the webhook is triggered)

I wonder if there's something wrong in your Python code. When you did the upgrade to v3.7.0, was it using a git pull, or a tarball?

**Reply by eramnes** [reply]
The upgrade was done via git pull.  I did not notice any errors during that process.

I did some other testing by calling the build pipeline manually in our CD system instead of using the webhook to call it from Netbox.  This led to some unexpected results.

In Netbox, there are actually two event triggers/webhooks that are used in the process.  The first one is the "specced" trigger that I've been referencing here.  In the CD pipeline that runs when that webhook triggers, the final step of that pipeline uses the Netbox Ansible plugin to set the VM status to "staged", which should trigger the second webhook.

When I manually ran the pipeline that would have run with the "specced" webhook, it completed successfully which means it was able to get to Netbox and collect the VM spec data.  The odd part was when the final step of changing the status to "staged" ran, as this successfully activated the second event trigger in Netbox and started the second CD pipeline automatically.  I was able to see the job in the Django queue and it processed successfully.

I created a new VM in Netbox using the web UI and set the status to "staged" instead of "specced" which would bypass the first build pipeline webhook and call the second phase using the other event trigger directly, but this had the same effect as setting the status to "specced" - nothing happened.

So I guess I'm not sure what's happening either as appears to use the event triggers, but only if the status isn't set in the web UI.  Not sure what the Ansible plugin might be doing on the backend that's different but maybe that helps narrow things down.  It appears that the Django queue is working as it did process the "staged" event as expected.

Thanks again for all the insight so far.

**Reply by eramnes** [reply]
I did some further testing this morning and some of my previous information is incorrect.  Using the "staged" webhook works whether the change is made via the web UI, API, or Ansible plugin.

We have a total of four event rules/webhooks configured in Netbox, and three of those don't work.  The only difference between the event rules is the value of the "status" condition, so I don't know why one works and the others don't.

I tried completely deleting the event rule and webhook for the "specced" status and recreating them with new names but they continue to experience the same issue.

I am considering rolling back to a snapshot from before the 3.7.0 upgrade just so I can get this back online.  I can keep the 3.7.0 one around as well for continued testing.

**Reply by markkuleinio** [reply]
FWIW, I tested changes with a prefix, with this condition in the event rule:

```
{
    "and": [
        {
            "attr": "status.value",
            "value": "deprecated"
        }
    ]
}
```

The event rule was triggered whenever I made a change in the prefix and the status of the prefix was Deprecated (when the change was saved). The event rule was not triggered when the prefix status was something else. Seems to work here (never used the conditions before).

**Reply by markkuleinio** [reply]
Also tried with a field choice
```
FIELD_CHOICES = {
    'ipam.Prefix.status+': (
        ("test", "Test", "red"),
    )
}
```
and changed the condition to use "test" instead of "deprecated". Still works as expected, event rule triggers only when status is Test.

**Reply by adparis99** [reply]
Just wanted to add in my experience, which seems to be somewhat similar to @eramnes. We also have a total of four webhooks that are used on a semi-regular basis, and none of them seem to work post-3.7. They are configured to execute when an interface with a given 802.1x type is updated (Access, Tagged, Tagged-All, etc). I've tried with and without conditions, but I haven't seen the Finished Job or Failed Job counters increase. To make matters stranger, at some point I decided to try the same webhook in the demo instance, which didn't seem to work at first. I can confirm now that it most certainly does work in the demo instance, both with and without conditions, but still does not work in my local instance. @candlerb, any ideas on what else to dig for?

Update: After some further testing, it looks like my problem is indeed almost identical to that of @eramnes. I was finally able to get one of our four webhooks to work, specifically one that targets interfaces with an 802.1x type of access. Unfortunately, webhooks aimed at interfaces with a different 802.1x type seem to be non-functional regardless of what I do. I also tested on the demo instance, but I couldn't replicate the behavior; webhooks seem to execute properly there regardless of the 802.1x type.

**Reply by markkuleinio** [reply]
> [...] none of them seem to work post-3.7. They are configured to execute when an interface with a given [802.1Q mode] is updated (Access, Tagged, Tagged-All, etc)

Do you mean that they all originally had a condition configured?

Do you have any webhooks without conditions? (If yes, are they working?)

> I was finally able to get one of our four webhooks to work

Any idea what was the change after which it started to work?

**Reply by candlerb** [reply]
> any ideas on what else to dig for?

Nope: if it was affecting me I would be sprinkling log or print statements around the code to try to understand where the webhook gets dropped.

**Reply by adparis99** [reply]
> Do you mean that they all originally had a condition configured?

Yes, that is correct. 

> Do you have any webhooks without conditions? (If yes, are they working?)

No we do not. I did try removing conditions from various webhooks in the troubleshooting process, but we didn't have any webhooks pre-upgrade without conditions.

> Any idea what was the change after which it started to work?

At the time I discovered there was an issue, I had all four webhooks enabled, each targeting dcim.interfaces. At some point I thought that maybe having all four was causing some sort of conflict, so I disabled all but the webhook that is currently working (that is, the one triggered for interfaces updated with an 802.1x type of access). That webhook continues to work now, even with the others enabled. So it's possible that disabling and re-enabling those had some kind of effect. It's also possible that the webhook that's working now was always working. I feel like I tested it and didn't see results, but I've done so many variations that it's possible I overlooked it. In any case, now I can shift my attention to figuring out why only that 802.1x type is triggering and not the others.

**Reply by DanSheps** [reply]
Is "802.1x type" a custom field?  802.1x is access control (nothing to do with access, tagged)

**Reply by adparis99** [reply]
@DanSheps Sorry, definitely meant 802.1Q mode.

**Reply by DanSheps** [reply]
Ah yeah, that makes sense!

**Comment by markkuleinio** [comment]
@eramnes @adparis99 What was the earlier NetBox version you had before upgrading to 3.7.0?

**Reply by eramnes** [reply]
We were on 3.6.5 before upgrading.  I did go ahead and roll back our Netbox instance to the 3.6.5 backup, and I can confirm that reverting fixed the issue I was seeing with the webhooks.  It's Saturday here, but I am planning to try the 3.7.0 upgrade again this coming week, however I'm going to delete all webhooks before the upgrade and then recreate them once on 3.7.0 to see if that makes a difference.

If doing an interim upgrade first before moving to 3.7.0 might help (e.g. 3.6.5 -> 3.6.9 or something) I can try that as well.  Also if anyone has any thoughts on where to put in some logging statements I would be willing to do that as well.

I have been wondering if somehow the Webhook model changes didn't quite complete successfully during the upgrade, and there are still pieces of the old model associated with the webhooks that don't work.  Just a guess as I'm not familiar with how the code really functions.  In any case, if removing the webhooks first and recreating them after works, that will be a solution for us.

**Reply by markkuleinio** [reply]
Ok, I just did a quick 3.6.5 install (using https://github.com/markkuleinio/netbox-freezes/blob/main/netbox-3.6.5.txt), added a working webhook (with a condition on prefix status "deprecated"), then upgraded to 3.7.0, still works. So there is something more complicated behind this.

**Reply by adparis99** [reply]
Honestly - I can't remember. It was either 3.6.7 or 3.6.8.

**Comment by adparis99** [comment]
@candlerb @markkuleinio I believe I've found a way to replicate the behavior I and @eramnes have been seeing in the NetBox demo instance. To put it succinctly, I suspect that when multiple event rules with different conditions are applied to a single model, only the first event rule is being evaluated (and executed, should the condition match). In my example, I'm using dcim.interface as the model, and the 802.1x mode as the conditions (access, tagged, and tagged-all).
Would you mind trying the steps below and note if the behavior matches my description?

In my testing, event rules do not trigger (or more specifically, a webhook is not generated) for tagged interfaces after following these steps, but they do trigger for access interfaces: 
1. Create two webhooks pointed toward a webhook receiver
2. Create two event rules
    * Configure the first rule to be enabled, associated with first of the two new webhooks, triggered by updates to dcim.interfaces, and with the condition below:
        * `{  "and": [ { "attr": "mode.value", "value": "access" }  ] }`
    * Configure the second rule the same, but associated with the second of the two new webhooks, and with the condition below:
        * `{  "and": [ { "attr": "mode.value", "value": "tagged" }  ] }`
3. Update an interface with an 802.1x mode of access, and then update an interface with an 802.1x mode of tagged
4. Check webhook receiver. There is a webhook based on the interface with mode access, but not with mode tagged

If I swap the conditions between the two rules - that is, I put the tagged condition on the first event rule that I created, and access on the second rule - the behavior swaps as well. I can then generate webhooks for tagged interfaces, but not for access.

**Reply by markkuleinio** [reply]
Good catch.

Issue #14827 opened, please comment there as well if you have something to add.

**Comment by kuhball** [comment]
We have encountered a maybe related issue. Recently upgraded our netbox prod instance from 3.6.6 to 3.7.1. Doing a bulk edit on 2 interfaces resulted in only one webhook. Expected are 2 webhooks, one for each interface.

Unfortunately I wasn't able to reproduce this with the following steps:

- netbox 3.6.6 with imported demo data
- create a webhook for interfaces pointing to http://127.0.0.1:9000
- try a bulk change -> 2 webhooks
- upgrade to 3.7.1
- try a bulk change also 2 webhooks

Importing our database into a fresh 3.7.1 docker netbox instance recreates this problem. Was also thinking this could be a problem with having multiple webhooks for the same object types (we currently have 3 webhooks for interface object types). But this was also not reproducible. 

The solution for us was to delete all the event rules and recreate them. Afterwards all bulk edits produced the expected webhooks for all edited objects. 

Posting this mostly if other users happen to have a similar problem.
